### PR TITLE
21 remove token 2

### LINKED ISSUE #21: Feature: Set weight of REMOVE_TOKEN to 0.5%
## User Story
**As a** controller, **I want to** remove tokens that don't match the pool's investment thesis, **so that** the pool matches my investment thesis.

## Acceptance Criteria
- [ ] when a call to `reindexTokens` does **not** include a token that is currently held by the pool, it can be assumed that the token should to be removed from the pool and it's desired end weight is set to 0.5%

### PR DESCRIPTION
**Pull request by BritikovKI**
Closes #21

### CODE REVIEW

File: pkg/pool-weighted/contracts/smart/IndexPool.sol [hunk 1]
```diff
@@ -240,10 +256,11 @@ contract IndexPool is BaseWeightedPool, ReentrancyGuard, IIndexPool {
     ) public override returns (uint256) {
         //cannot swap out uninitialized token
         _require(minBalances[swapRequest.tokenOut] == 0, Errors.UNINITIALIZED_TOKEN);
+        require(_tokenState[swapRequest.tokenIn].decodeUint5(_REMOVE_FLAG_OFFSET) != _REMOVE_FLAG, "Removed token");
```
Comment (Zitzak) [inline pkg/pool-weighted/contracts/smart/IndexPool.sol:259]: Maybe add the error message to the `BalancerErrors.sol` so they way we show errors will be uniform

Comment (BritikovKI) [reply pkg/pool-weighted/contracts/smart/IndexPool.sol]: Fixed

File: pkg/pool-weighted/contracts/smart/IndexPool.sol [hunk 2]
```diff
@@ -50,8 +50,8 @@ contract IndexPool is BaseWeightedPool, ReentrancyGuard, IIndexPool {
 
     // Store scaling factor and start/end weights for each token
     // Mapping should be more efficient than trying to compress it further
-    // [ 123 bits| 32 bits        |   5 bits |  32 bits   |   64 bits    ]
-    // [ unused  | target weights | decimals | end weight | start weight ]
+    // [ 118 bits | 5 bits | 32 bits        |   5 bits |  32 bits   |   64 bits    ]
```
[the remark this answers is not in the document]
Comment (Zitzak) [reply pkg/pool-weighted/contracts/smart/IndexPool.sol:53]: Oke 😂

**Review by Zitzak — APPROVED** [review]
lgtm

**Review by fabianschu — COMMENTED** [review]
After going through the code, I'm wondering if a nested loop wouldn't be the better approach here after all.. 

I try to recap my understanding:
If I get that right, now we're iterating over the reindex tokens and store a flag for each token ("SAVE"-flag). Then we iterate over the tokens in the pool and for each pool token check if it has this SAVE-flag. If it hasn't we know it needs to be removed, so we store a REMOVE flag. If it has, we restore the NORMAL flag.

This means that for each reindex call, depending on how many tokens are being removed, just for the `assembleReindexParams` we write to storage up to `2 * amountOfReindexParams` times and that is before we are actually writing to storage the newly updated token state in `startGradualWeightChange`.

Example: We have 4 tokens in the pool and one should be removed. In `assembleReindexParams` we will write 3 + 4 times to storage, which is already at least 7 * 20k gas just by doing that. (later on we write another 4 times to storage for updating the actual new weights, etc.).

Tbh, I believe doing a bunch of looping with bounded loops, but all in memory, is much more gas efficient in the end. Wdyt?

**Comment by fabianschu** [comment]
Maybe we could first check if there are any new tokens being added (which is kinda cheap) and then if we know that this is not the case we just skip this whole removing tokens hell by comparing the sizes of the input params and the tokens in the pool, so that we only incur these high costs if there is really a token being removed?

**Comment by BritikovKI** [comment]
> After going through the code, I'm wondering if a nested loop wouldn't be the better approach here after all..
> 
> I try to recap my understanding: If I get that right, now we're iterating over the reindex tokens and store a flag for each token ("SAVE"-flag). Then we iterate over the tokens in the pool and for each pool token check if it has this SAVE-flag. If it hasn't we know it needs to be removed, so we store a REMOVE flag. If it has, we restore the NORMAL flag.
> 
> This means that for each reindex call, depending on how many tokens are being removed, just in the context of the `assembleReindexParams` function, we write to storage up to `2 * amountOfReindexParams` times and that is before we are actually writing to storage the newly updated token state in `startGradualWeightChange`.
> 
> Example: We have 4 tokens in the pool and one should be removed. In `assembleReindexParams` we will write 3 + 4 times to storage, which is already at least 7 * 20k gas just by doing that. (later on we write another 4 times to storage for updating the actual new weights, etc.).
> 
> Tbh, I believe doing a bunch of looping with bounded loops, but all in memory, is much more gas efficient in the end. Wdyt?

A very good question, I love it tbh :)
To be sure we need to test it, but here is theory:

> Persistent storage on opcode (SSTORE) is extremely expensive. The current cost per one 32-byte word is 20,000 Gas (~ 5 cents at 10 Gwei gas price and $250 per ETH) for the first time a particular slot is set, and 5,000 Gas for each subsequent modification.

That's an info from [here](https://medium.com/bandprotocol/solidity-102-1-keeping-gas-cost-under-control-ae95b835807f). If we are going to do the writing into state it is going to take us 250000 gas at most(Because we are writing in the already allocated memory). If we're going to do iteration inside of iteration it is going to lead to complexity O(n^2) where we're going to read from the storage and execute some commands from inside of the loop, let's suppose we are going to have 1000 gas price per loop iteration(we need to increment integer, get the variable, do the comparison), and we have 50 tokens, this means that we're going to have 50*50 iterations, which is equal to 2500(Actually it will be smth closer to 1250 because of random token placing, and if all tokens remain intact we're not going to loose anything ), which leads us to smth like 1250000 gas, it is speculation tbh, but even if we have 100 gas cost per check, it will still be 125000, which is not that much less then 250000

**Comment by BritikovKI** [comment]
> Maybe we could first check if there are any new tokens being added (which is kinda cheap) and then if we know that this is not the case we just skip this whole removing tokens hell by comparing the sizes of the input params and the tokens in the pool, so that we only incur these high costs if there is really a token being removed?

There is no sense to call reindex without any tokens being added/removed, and we cannot be sure that there are no tokens being removed, until we check all the input tokens...

File: pkg/pool-weighted/contracts/smart/IndexPool.sol [hunk 3]
```diff
@@ -64,6 +64,11 @@ contract IndexPool is BaseWeightedPool, ReentrancyGuard, IIndexPool {
     uint256 private constant _DECIMAL_DIFF_OFFSET = 96;
     uint256 private constant _UNINITIALIZED_OFFSET = 101;
     uint256 private constant _NEW_TOKEN_TARGET_WEIGHT_OFFSET = 106;
+    uint256 private constant _REMOVE_FLAG_OFFSET = 138;
+
+    uint256 private constant _NORMAL_FLAG = 0;
+    uint256 private constant _SAVE_FLAG = 1;
+    uint256 private constant _REMOVE_FLAG = 2;
```
[the remark this answers is not in the document]
Comment (BritikovKI) [reply pkg/pool-weighted/contracts/smart/IndexPool.sol:71]: NORMAL_FLAG is just untouched tokens, with it tokens can be set to REMOVE, SAVE flag protects tokens for 1 iteration, with it tokens cannot be set to REMOVE, REMOVE flag shows that tokens are going to be removed(never going to change)

File: pkg/pool-weighted/contracts/utils/IndexPoolUtils.sol [hunk 4]
```diff
@@ -132,35 +137,68 @@ library IndexPoolUtils {
...
+    /// @return initialFixedWeights Weights that are fixed and that the other non-fixed weights need to be adjusted for.
+    /// @return finalFixedWeights Weights that are fixed and that the other non-fixed weights need to be adjusted for.
     /// @return newTokenTargetWeights contains only the final target weights for uninitialized tokens (else zero).
-    /// @return existingTokens contains only the existing pool token addresses (else zero).
+    /// @return newDesiredWeights contains only the final desired weights for tokens.
     /// @return newTokens contains only the new pool token addresses (no zeros!).
+    /// @return finalTokens contains final pool token addresses (no zeros!).
     function assembleReindexParams(
         IERC20[] memory tokens,
+        IERC20[] memory oldTokens,
```
[the remark this answers is not in the document]
Comment (BritikovKI) [reply pkg/pool-weighted/contracts/utils/IndexPoolUtils.sol:149]: Fixed

**Review by fabianschu — APPROVED** [review]

**Pull request merged**